src/kaleidoscope/hid.cpp

```diff
@@ -51,93 +51,225 @@
+// for this report.  This is set when a keypress is first detected and cleared 
+// after the report is sent. If multiple keys are toggled on during a single
+// cycle, this contains the most recently handled one. 
+
+static uint8_t last_keycode_toggled_on{0};
+
+void resetModifierTracking(void) {
+  last_keycode_toggled_on = 0;
+  requested_modifiers = 0;
 }
 
-void pressRawKey(Key mappedKey) {
-  WITH_BOOTKEYBOARD_PROTOCOL {
-    BootKeyboard.press(mappedKey.keyCode);
-    return;
-  }
+// isModifierKey takes a Key and returns true if the key is a 
+// keyboard key corresponding to a modifier like Control, Alt or Shift
+// TODO: This function should be lifted to the Kaleidoscope core, somewhere.
+
+bool isModifierKey(Key key) {
+  // If it's not a keyboard key, return false
+  if (key.flags & (SYNTHETIC | RESERVED)) return false;
+
+  return (key.keyCode >= HID_KEYBOARD_FIRST_MODIFIER &&
+          key.keyCode <= HID_KEYBOARD_LAST_MODIFIER);
+}
+
+// requestModifiers takes a bitmap of modifiers that might apply
+// to the next USB HID report and adds them to a bitmap of all such modifiers.
 
-  Keyboard.press(mappedKey.keyCode);
+void requestModifiers(byte flags) {
+  requested_modifiers |= flags;
 }
 
-void pressKey(Key mappedKey) {
-  if (mappedKey.flags & SHIFT_HELD) {
+// pressModifiers takes a bitmap of modifier keys that must be included in
+// the upcoming USB HID report and passes them through to KeyboardioHID
+// immediately
+
+void pressModifiers(byte flags) {
+  if (flags & SHIFT_HELD) {
     pressRawKey(Key_LeftShift);
   }
-  if (mappedKey.flags & CTRL_HELD) {
+  if (flags & CTRL_HELD) {
     pressRawKey(Key_LeftControl);
   }
-  if (mappedKey.flags & LALT_HELD) {
+  if (flags & LALT_HELD) {
     pressRawKey(Key_LeftAlt);
   }
-  if (mappedKey.flags & RALT_HELD) {
+  if (flags & RALT_HELD) {
     pressRawKey(Key_RightAlt);
   }
-  if (mappedKey.flags & GUI_HELD) {
+  if (flags & GUI_HELD) {
     pressRawKey(Key_LeftGui);
   }
+}
+
+}  // namespace
+
+
+
+void initializeKeyboard() {
+  Keyboard.begin();
+  WITH_BOOTKEYBOARD {
+    BootKeyboard.begin();
+  }
+}
+
+// pressKey takes a Key, as well as optional boolean 'toggledOn' which defaults
+// to 'true'
+
+// If toggled_on is not set to false, this routine adds the modifier flags on
+// this key to the bitmask of modifiers that are allowed to be added to the
+// upcoming report. We do this so that when we roll over from a key with a
+// modifier flag to one without it, that modifier flag won't affect the new
+// keypress.
+
+// If the key we're processing is a modifier key, any modifier flags attached to
+// it are added directly to the report along with the modifier from its keycode
+// byte.
+//
+// (A 'modifier key' is one of the eight modifier keys defined by the HID
+// standard: left and right variants of Control, Shift, Alt, and GUI.)
+
+// Eventually it calls pressRawKey.
+
+void pressKey(Key pressed_key, boolean toggled_on) {
```

Comment:
```cpp
void pressKey(Key pressed_key, boolean toggled_on=true) {
```

Reply:
(See my comment on the other related change. We set this in the API definition)